lib/index.js

```diff
@@ -50,87 +51,92 @@ function getMinimumLimit(limit1, limit2) {
   return limit2;
 }
 
-function setupRateLimitEventExt(server, options) {
-  const event = options.event;
-  const extractKey = options.extractKey;
-  const onError = options.onError;
-
-  const extractKeyAndTakeToken = function(limitd, request, reply, type) {
-    extractKey(request, reply, (err, key) =>{
-      if (err) { return reply(err); }
-
-      if (!limitd) {
-        // limitd is not connected, do not fail!
-        return reply.continue();
-      }
-
-      limitd.take(type, key, (err, currentLimitResponse) => {
-        if (err){
-          if (onError) { return onError(err, reply); }
-          // by default we don't fail if limitd is unavailable
-          return reply.continue();
-        }
-
-        const oldMinimumLimitResponse = request.plugins.patova && request.plugins.patova.limit
-        const newMinimumLimitResponse = getMinimumLimit(currentLimitResponse, oldMinimumLimitResponse)
+function buildRateLimitEvent({ extractKey, limitd, onError, type }) {
+  async function getType(request) {
+    if (typeof type !== 'function') {
+      return type;
+    }
 
-        request.plugins.patova = request.plugins.patova || {};
-        request.plugins.patova.limit = newMinimumLimitResponse;
+    try {
+      return await type(request);
+    } catch (err) {
+      throw Boom.wrap(err, 500, 'cannot get bucket type');
+    }
+  }
 
-        if (newMinimumLimitResponse.conformant) {
-          // We continue only if the request is conformat so far
-          return reply.continue();
+  function takeAsync(type, key) {
```

Comment:
utils.promisify? https://nodejs.org/dist/latest-v8.x/docs/api/util.html#util_util_promisify_original?

Reply:
That was my first tough, but when i do something like:

```js
  const takeAsync = util.promisify(limitd.take);
```

It's fails like this:
```
TypeError: this._takeOrWait is not a function
    at LimitdClient.take (/home/jose/Code/auth0/update-hapi-17/patova/node_modules/limitd-client/client.js:406:15)
    at internal/util.js:230:26
    at onRateLimit (/home/jose/Code/auth0/update-hapi-17/patova/lib/index.js:99:36)
    at <anonymous>
```

It's clearing loosing the object context; i could try to evaluate `takeAsync` in `limitd` context, but i think that is clear just to create that function.

Reply:
what about `const takeAsync = util.promisify(limitd.take.bind(limitd));` I think it is a really common pattern and I think it is clear. The general advice from bluebird people is to try to limit building promises by hand as much as possible (http://bluebirdjs.com/docs/anti-patterns.html), the same from Q guys, and the reason is that it is easy to miss edge cases or to loose guaranties (like realising zalgo :P https://oren.github.io/blog/zalgo.html)

Reply:
No objections, but just for the record, I think is perfectly safe to use `new Promise()`. the only valid objection from bluebird guys is that 'you are reinventing the wheel', since the other one (throwing an sync exception) does not apply. Doing `new Promise(() => { throw new Error() })` works perfectly fine, and the exception will not be reported synchronously. But I'm ok if for style/conciseness reasons we prefer the `promisify`.